### PR TITLE
[WIP] Use fsspec to manage access to resources and caching

### PR DESCRIPTION
**Pull request by gmaze**
The goal of this PR is to implement [fsspec](https://filesystem-spec.readthedocs.io) as file system in order to cache local files.
This PR should be closing #12, #15 and #17.

### CODE REVIEW

**Comment by gmaze** [comment]
Inspiration:
- https://xmitgcm.readthedocs.io/en/latest/llcreader.html
- https://github.com/MITgcm/xmitgcm/blob/master/xmitgcm/llcreader/stores.py

DESIGN for FTP (local or online) fetchers
=========================================

In gdacreader/stores.py:

```python
class BaseStore
    def __init__(self, fs, base_path='/')
    #
    # base_path could be:
    #   'ftp://ftp.ifremer.fr/ifremer/argo/' # for online store
    #   '/home/' # for offline/local store
        self.fs = fs
        self.base_path = base_path

    def _dac(self, wmo)
    # Return DAC name for this WMO
    # Needed for path construction
    # eg: 'coriolis' for wmo=1900065

    def _fname(self, wmo, cyc, varname)
    # Return GDAC file name where to find a specific variable
    # eg: 'profiles/D1900065_003.nc' for wmo=1900065, varname='PSAL' and cyc=3
    # eg: 'D1900065_prof.nc' for wmo=1900065, varname='PSAL' and cyc=None
    # This should allow us to manage core, biogeochemical and trajectory measurements (all in different files)

    def _directory(self, dac, wmo, cyc)
    # Return GDAC base directory to access a file
    # eg: 'ftp://ftp.ifremer.fr/ifremer/argo/dac/coriolis/1900065/' for dac='coriolis', wmo=1900065 and cyc=None
    # eg: 'ftp://ftp.ifremer.fr/ifremer/argo/dac/coriolis/1900065/profiles/' for dac='coriolis', wmo=1900065 and cyc=3

    def _full_path(self, dac, wmo, cyc)
    # Return full GDAC path to a file
    # eg: 'ftp://ftp.ifremer.fr/ifremer/argo/dac/coriolis/1900065/1900065_prof.nc' for dac='coriolis', wmo=1900065 and cyc=None
    # eg: 'ftp://ftp.ifremer.fr/ifremer/argo/dac/coriolis/1900065/profiles/D1900065_003.nc' for dac='coriolis', wmo=1900065 and cyc=3

    def get_fs_and_full_path(self, wmo, cyc, varname)
    # Return references to a filesystem and path within it for a specific WMO, cycle number and variable
    #   Returns
    #   -------
    #   fs : fsspec.AbstractFileSystem
    #       The filesytem where the file can be found
    #   path : str
    #       The path to open
        return self.fs, self._full_path(wmo, cyc, varname)

    def open_data_file(self, wmo, cyc, varname)
    # Open the file for a specific WMO, cycle number and variable
    #   Parameters
    #   ----------
    #   wmo : int
    #   cyc : int
    #   varname : str
    #
    #   Returns
    #   -------
    #    fobj : file-like object
        fs, path = self.get_fs_and_full_path( wmo, cyc, varname)
        return fs.open(path)

```

In gdacreader/readers.py:
```python
def _make_ftp_filesystem():
    import fsspec
    from fsspec.implementations.ftp import FTPFileSystem
    return FTPFileSystem()

def _make_local_filesystem():
    import fsspec
    from fsspec.implementations.local import LocalFileSystem
    return LocalFileSystem()

class BaseFTP()
    def __init__(self, store)

    def _filepathpattern

    def get_dataset(self, wmo, cyc)
    # Create an xarray Dataset object for this profile.

class OnlineFTP(BaseFTP)
    def __init__(self)
        fs = _make_ftp_filesystem()
        base_path = 'ftp://ftp.ifremer.fr/ifremer/argo'
        store = gdacreader.BaseStore(fs, base_path=base_path)
        super().__init__()

class LocalFTP(BaseFTP)
    def __init__(self, base_path='.')
        fs = _make_local_filesystem()
        store = gdacreader.BaseStore(fs, base_path=base_path)
        super().__init__()
```

USAGE
=====
From data/index fetchers, we would be using:

```python
from argopy import gdacreader

store = gdacreader.OnlineFTP()
ds = store.get_dataset(wmo=690123, cyc=12)

store = gdacreader.LocalFTP()
ds = store.get_dataset(wmo=690123, cyc=12)
```

QUESTIONS
=========
Need to clarify how to manage a collection of files corresponding to a user selection:
1. is it with a collection of stores ?
1. is it with multiple request to the store ?

It's probably the 2nd one, but I still don't figure how to implement this

Another question is how to design this to also manage access to the erddap. But we may want to wait for erddapy to implement fsspec.

**Comment by gmaze** [comment]
At this point, I implemented a file system based on fsspec in the erddap fetcher that manage both cache and no cache options. All CI testings is completed and updated.

All requests to the Erddap are now delegated to an internal "fs" instance that is created when the fetcher is initialized by argopy.fetchers.ArgoDataFetcher. In the erddap.py ErddapArgoDataFetcher class we have:
```python
        self.fs = onlinestore(cache=cache, cachedir=cachedir)
```

In argopy.utilities I created the ``onlinestore`` class that is a wrapper around [fsspec http file store](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.implementations.http.HTTPFileSystem). I used a wrapper to better handle specific errors from requests to the erddap.

**Comment by gmaze** [comment]
I forgot to mention that the ``onlinestore`` should be able to replace the former ``urlopen`` and also to work with the upcoming argovis fetcher, also based on web requests

**Comment by gmaze** [comment]
I have implemented a local file storage system using [fsspec LocalFileSystem](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.implementations.local.LocalFileSystem) for the local_ftp data fetcher.

I have difficulties with implementing it with the local_ftp index fetcher because the caching logic is different from @quai20 code. There are 2 levels of caching and I suspect there is room for improvement here.

**Comment by gmaze** [comment]
Lot of work since yesterday evening on this.
It was quite complex to manage the localftp index fetcher because they are 2 levels of cache needed, one for the full index and one for the search request.
I ended up implementing a [fsspec LocalFileSystem](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.implementations.local.LocalFileSystem) for the index and a [fsspec MemoryFileSystem](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.implementations.memory.MemoryFileSystem) for the search. This seems to work quite nicely.
I also modified the internal factoring of the index fetcher, the ``filter_index`` method now consume an index file pointer and returns in-memory text with the csv rows. This allows to get rid of file manipulation in too many places and seemingless integration with fsspec in the fetcher.
I also improved the index search method to make it much faster and able to look for a float cycle numbers (allowing to implement the 'profile' entry point with localftp).

Internal design for this is still not satisfactory though: fsspec instance are still created in localftp.py and not part of an upper level utility class. This is my next step.

**Comment by gmaze** [comment]
A lot of internal refactoring:

- The localftp index fetcher now uses only utility functions and classes to manage access to the index and searches, with cache or not.
- Introduces an ``indexstore`` that should be able to manage local and online searches on the Argo index.
- Searches are based on a new ``index_filter_wmo`` that inherits from a ``index_filter_proto``. These work in memory and will be reusable for online and offline index stores.
- All file/resource access/management utilities are now under the sub-packages ``stores``.

File: argopy/stores/argo_index.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+            csv chunk matching the request, as a string. Or None
+            """
+
+            def search_this(this_line):
+                # return np.any([re.search("%0.3d.nc" % c, this_line.split(',')[0]) for c in cyc])
+                return np.any(["%0.3d.nc" % c in this_line for c in cyc])
+                if np.all(cyc >= 1000):
+                    def search_this(this_line):
+                        # return np.any([re.search("%0.4d.nc" % c, this_line.split(',')[0]) for c in cyc])
+                        return np.any(["%0.4d.nc" % c in this_line for c in cyc])
```
Comment (quai20) [inline argopy/stores/argo_index.py:181]: Unreachable "if" and function defined twice

Comment (gmaze) [reply argopy/stores/argo_index.py:181]: Yes I don't know what to do about this
The point is to make sure that we can handle float with cycle numbers larger than 1000, and for which the file name will be different, any idea ?

Comment (gmaze) [reply argopy/stores/argo_index.py:181]: The issue is that for a float with more than 1000 cycle, the cycle number 1 name will be: _0001.nc and not _001.nc

Comment (gmaze) [reply argopy/stores/argo_index.py:181]: oh shoot, now I get it !!! the return comes too early !

Comment (quai20) [reply argopy/stores/argo_index.py:181]: yep

Comment (gmaze) [reply argopy/stores/argo_index.py:181]: Here is a fix:

```python
            if np.all(cyc >= 1000):
                def search_this(this_line):
                    # return np.any([re.search("%0.3d.nc" % c, this_line.split(',')[0]) for c in cyc])
                    return np.any(["%0.3d.nc" % c in this_line for c in cyc])
            else:
                def search_this(this_line):
                    # return np.any([re.search("%0.4d.nc" % c, this_line.split(',')[0]) for c in cyc])
                    return np.any(["%0.4d.nc" % c in this_line for c in cyc])
```

File: argopy/stores/argo_index.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+            # Then look for the profile:
+            if results:
+                def search_this(this_line):
+                    # return np.any([re.search("%0.3d.nc" % c, this_line.split(',')[0]) for c in cyc])
+                    return np.any(["%0.3d.nc" % c in this_line for c in cyc])
+                if np.all(cyc >= 1000):
+                    def search_this(this_line):
+                        # return np.any([re.search("%0.4d.nc" % c, this_line.split(',')[0]) for c in cyc])
+                        return np.any(["%0.4d.nc" % c in this_line for c in cyc])
```
Comment (quai20) [inline argopy/stores/argo_index.py:232]: same here

Comment (gmaze) [reply argopy/stores/argo_index.py:232]: yep

Comment (gmaze) [reply argopy/stores/argo_index.py:232]: Ok I fixed this in the 2 places

**Comment by gmaze** [comment]
Ok, so I will add a cache clearing system and we're almost there !

**Pull request merged**